File: sdk/java/src/main/java/com/anyvali/ValidationIssue.java
package com.anyvali;

import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * A single validation issue.
 */
public record ValidationIssue(
        String code,
        String message,
        List<Object> path,
        Object expected,
        Object received,
        Map<String, Object> meta
) {
    public ValidationIssue {
        path = path != null ? List.copyOf(path) : List.of();
        meta = meta != null ? Map.copyOf(meta) : null;
    }

    public ValidationIssue(String code, String message, List<Object> path,
                           Object expected, Object received) {
        this(code, message, path, expected, received, null);
    }

    public ValidationIssue(String code, String message, List<Object> path) {
        this(code, message, path, null, null, null);
    }
}

File: sdk/kotlin/src/main/kotlin/com/anyvali/ValidationIssue.kt
package com.anyvali

data class ValidationIssue(
    val code: String,
    val message: String = "",
    val path: List<Any> = emptyList(),
    val expected: String = "",
    val received: String = "",
    val meta: Map<String, Any?> = emptyMap()
)
